src/transformStateWithClones.js

```diff
@@ -6,8 +6,66 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+
+const state = {
+  foo: 'bar',
+  name: 'Jim',
+  another: 'one',
+};
+
+function transformStateWithClones(currentState, actions) {
+  const history = [];
+  let clonedState = Object.assign({}, currentState);
+
+  actions.forEach((obj) => {
+    if (obj.type === 'clear') {
+      clonedState = {};
+      history.push({});
+    }
+
+    if (obj.type === 'addProperties') {
+      clonedState = { ...clonedState, ...obj.extraData };
+      history.push({ ...clonedState });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const key of obj.keysToRemove) {
+        delete clonedState[key];
+      }
+
+      history.push({ ...clonedState });
```

Comment:
Again, avoid pushing the `clonedState` inside the 'removeProperties' case. Push the `clonedState` to the history array once at the end of the loop cycle. [DONT REPEAT YOURSELF]